optional-deps/src/index.rs

```diff
@@ -0,0 +1,146 @@
+}
+
+/// Dependencies include mandatory dependencies and optional dependencies.
+/// Optional dependencies are identified by an option called a "feature".
+pub struct Deps {
+    /// The regular, mandatory dependencies.
+    pub mandatory: Map<Package, Dep>,
+    /// The optional, feature-gated dependencies.
+    pub optional: Map<Feature, Map<Package, Dep>>,
+}
+
+/// A dependency is specified with a range, and with a set of activated features.
+pub struct Dep {
+    /// The range dependended upon.
+    pub range: Range<Version>,
+    /// The activated features for that dependency.
+    pub features: Set<Feature>,
+}
+
+impl Default for Deps {
+    fn default() -> Self {
+        Self {
+            mandatory: Map::default(),
+            optional: Map::default(),
+        }
+    }
+}
+
+impl Index {
+    /// Empty new index.
+    pub fn new() -> Self {
+        Self {
+            packages: Map::default(),
+        }
+    }
+
+    /// List existing versions for a given package with newest versions first.
+    pub fn available_versions(&self, package: &Package) -> impl Iterator<Item = &Version> {
+        self.packages
+            .get(package)
+            .map(|k| k.keys())
+            .into_iter()
+            .flatten()
+            .rev()
+    }
+
+    /// Register a package and its mandatory dependencies in the index.
+    pub fn add_deps<R: RangeBounds<u32>>(
+        &mut self,
+        package: &str,
+        version: u32,
+        mandatory_deps: &[(&str, R, &[&str])],
+    ) {
+        let deps = self
+            .packages
+            .entry(package.to_string())
+            .or_default()
+            .entry(version.into())
+            .or_default();
+        for (p, r, features) in mandatory_deps {
+            let dep = Dep {
+                range: range_from_bounds(r),
+                features: features.iter().map(|s| s.to_string()).collect(),
+            };
+            deps.mandatory.insert(String::from(*p), dep);
+        }
+    }
+
+    /// Register a feature and its associated dependencies for a given package in the index.
+    pub fn add_feature<R: RangeBounds<u32>>(
+        &mut self,
+        package: &str,
+        version: u32,
+        feature: &str,
+        optional_deps: &[(&str, R, &[&str])],
+    ) {
+        let deps = self
+            .packages
+            .entry(package.to_string())
+            .or_default()
+            .entry(version.into())
+            .or_default()
+            .optional
+            .entry(feature.to_string())
+            .or_default();
+        for (p, r, features) in optional_deps {
+            let dep = Dep {
+                range: range_from_bounds(r),
+                features: features.iter().map(|s| s.to_string()).collect(),
+            };
+            deps.insert(String::from(*p), dep);
+        }
+    }
+}
+
+/// Convert a range bounds into pubgrub Range type.
+fn range_from_bounds<R: RangeBounds<u32>>(bounds: &R) -> Range<Version> {
```

Comment:
I love this interface! Should this be in PubGrub (as `impl From<RangeBounds<V>> for Range<Version>`) instead of a helper function that everyone needs to copy?

Reply:
I suppose we could? I practice I think the more interesting impl would be with `Into<V>`. Like here we have `u32` into `NumberVersion` or for semantic versions, in the branch where I explored the usage of `Interval`, I defined

```rust
// Ease the creation of SemanticInterval from classic ranges: (2, 0, 0)..(2, 5, 0)
impl<RB: RangeBounds<(u32, u32, u32)>> From<RB> for SemanticInterval {
    fn from(range: RB) -> Self {
        Self::new(
            map_bound(SemanticVersion::from, owned_bound(range.start_bound())),
            map_bound(SemanticVersion::from, owned_bound(range.end_bound())),
        )
    }
}
```

I'm not sure why I did that and not the `Into<V>` one, maybe the compiler was not happy, maybe I just didn't for no specific reason.